### PR TITLE
[JBPM-9457] Springboot tests for Kafka Emitter

### PR DESCRIPTION
**Pull request by gmunozfe**
**JIRA**: [JBPM-9457](https://issues.redhat.com/browse/JBPM-9457)
Added springboot integration tests for Kafka emitter.

**referenced Pull Requests**: 

* https://github.com/kiegroup/jbpm/pull/1803

### CODE REVIEW

**Review by fjtirado — APPROVED** [review]
Wonderful job Gonzalo, thank you very much for your support on thi

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+public class KafkaEmitterBrokerDownTest extends KafkaFixture {
+    
+    @Autowired
+    protected DeploymentService deploymentService;
+
+    @Autowired
+    protected ProcessService processService;
+    
+    @Autowired
+    protected CaseService caseService;
```
Comment (afalhambra) [inline kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java:59]: not used in the test
```suggestion
```

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    @After
+    public void cleanup() {
+        cleanup(deploymentService);
+        System.clearProperty("org.kie.jbpm.event.emitters.kafka.max.block.ms");
+    }
+    
+    @Test(timeout = 30000)
+    public void testEmitterNoKafkaBrokerTimeoutException() throws Exception {
```
Comment (afalhambra) [inline kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java:76]:
```suggestion
    public void testEmitterNoKafkaBrokerTimeoutException() {
```

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        Optional<ILoggingEvent> logEvent = getLog(listAppender);
+        assertEquals(TimeoutException.class.getCanonicalName(), logEvent.get().getThrowableProxy().getClassName());
+        
+        assertNotNull(processInstanceId);
+        assertTrue(processInstanceId > 0);
+
+        processService.abortProcessInstance(processInstanceId);
+        
+        ProcessInstance pi = processService.getProcessInstance(processInstanceId);
+        assertNull(pi);
```
Comment (afalhambra) [inline kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java:92]: wouldn't it be worth checking that KafkaEmitter is able to send messages to the topic once the timeout is removed (or increased)? - wdyt?

**Review by afalhambra — COMMENTED** [review]
Awesome job @gmunozfe! Just a few comments.

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
+KIE Server with all capabilities
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/README.md:1]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/main/resources/swagger.properties [hunk 5]
```diff
@@ -0,0 +1,4 @@
+title = Spring Boot KIE Server Application
+description = This sample project demonstrates how to use KIE Server services in SpringBoot
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/main/resources/swagger.properties:2]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    }
+    
+    protected CaseFile caseFile(int valueSize) {
+        Map<String, Object> data = new HashMap<>();
+        data.put(VAR_KEY, VAR_VALUE);
+        data.put(LARGE_VAR, RandomStringUtils.random(valueSize));
+        return CaseFile.builder().data(data).build();
+    }
+    
+    protected Map<String, OrganizationalEntity> roleAssigments() {
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java:186]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        assertTrue(event.get("source").toString().contains("/process/"+artifactId));
+        assertTrue(event.get("data") instanceof Map);
+        Map<String, Object> data = (Map<String, Object>) event.get("data");
+        assertTrue(data.get("compositeId").toString().contains("SpringBoot_"));
+        assertEquals("org.kie.server.testing:"+artifactId+":1.0.0", data.get("containerId"));
+        assertNotNull(data.get("id"));
+        assertEquals(-1, data.get("parentId"));
+        assertTrue(data.get(VARIABLES) instanceof Map);
+        Map<String,Object> variables = (Map<String,Object>) data.get(VARIABLES);
+        assertTrue(!variables.isEmpty());
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java:242]: good point, thanks!

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/resources/kjars/UserTaskCase/pom.xml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.kie.server.testing</groupId>
+    <artifactId>common-parent</artifactId>
+    <version>1.0.0.Final</version>
+  </parent>
+
+  <artifactId>UserTaskCase</artifactId>
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/resources/kjars/UserTaskCase/pom.xml:12]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    
+    @AfterClass
+    public static void teardown() {
+        kafka.stop();
+        System.clearProperty("org.kie.jbpm.event.emitters.kafka.boopstrap.servers");
+        System.clearProperty("org.kie.jbpm.event.emitters.kafka.client.id");
+    }
+
+    @Test(timeout = 30000)
+    public void testKafkaEmitterProcessStartAndAbort() throws Exception {
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java:108]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    
+    @Test(timeout = 30000)
+    public void testKafkaEmitterProcessStartAndWorkOnUserTasks() throws IOException {
+        deploymentId = setup(deploymentService, EVALUATION_PROCESS_ID);
+        
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("employee", JOHN);
+        parameters.put("reason", "autoevaluation");
+        long processInstanceId = processService.startProcess(deploymentId, EVALUATION_PROCESS_ID, parameters);
+        assertNotNull(processInstanceId);
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java:134]: good catch!

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        
+        List<TaskSummary> tasks = runtimeDataService.getTasksAssignedAsPotentialOwner(JOHN, new QueryFilter());
+        assertEquals(1, tasks.size());
+        
+        userTaskService.completeAutoProgress(tasks.get(0).getId(), JOHN, new HashMap<>());
+        
+        tasks = runtimeDataService.getTasksAssignedAsPotentialOwner(JOHN, new QueryFilter());
+        assertEquals(2, tasks.size());
+        
+        tasks.stream().forEach(t -> userTaskService.completeAutoProgress(t.getId(), JOHN, new HashMap<>()));
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java:145]: right, nice observation!

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        processService.abortProcessInstance(processInstanceId);
+        
+        ProcessInstance pi = processService.getProcessInstance(processInstanceId);
+        assertNull(pi);
+                
+        consumeAndAssertRecords(PROCESSES_TOPIC, PROCESS_TYPE, STATE_ABORTED, 1);
+    }
+    
+    @Test(timeout = 30000)
+    public void testKafkaEmitterProcessStartAndWorkOnUserTasks() throws IOException {
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java:127]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        assertEquals(4, records.count());
+        Map<String, Long> processRecordsByStatus =  groupRecordsByField(records, "state");
+        
+        assertEquals(3, processRecordsByStatus.get("1").intValue()); //Active
+        assertEquals(1, processRecordsByStatus.get("2").intValue()); //Completed
+        
+    }
+    
+    @Test(timeout = 30000)
+    public void testKafkaEmitterCaseStartAndAbort() throws Exception {
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java:169]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+        caseService.cancelCase(caseId);
+        
+        consumeAndAssertRecords(CASES_TOPIC, CASE_TYPE, STATE_ABORTED, 1);
+        
+        ConsumerRecords<String, byte[]>  records = consumeMessages(TASKS_TOPIC);
+        assertEquals(2, records.count());
+    }
+    
+    @Test(timeout = 30000)
+    public void testEmitterRecordTooLargeException() throws Exception {
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterHappyPathTest.java:190]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/ProxyAwareKafkaContainer.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+public class ProxyAwareKafkaContainer extends KafkaContainer {
+
+    private String host;
+    
+    public void setHost(String host) {
+        this.host = host;
+    }
+    
+    public String getBootstrapServers() {
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/ProxyAwareKafkaContainer.java:29]: done

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java (hunk 1 above)
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java:59]: removed

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java (hunk 2 above)
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java:76]: removed

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java (hunk 3 above)
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/KafkaEmitterBrokerDownTest.java:92]: Not in this test, as there's no testcontainer kafka broker

**Review by afalhambra — APPROVED** [review]
Awesome job @gmunozfe!

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/ProxyAwareKafkaEmitterTest.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+    protected static final int TOXY_PROXY_PORT = Integer.parseInt(System.getProperty("toxiproxy.port"));
+
+    public static Network network = Network.newNetwork();
+    
+    public static ProxyAwareKafkaContainer kafka = (ProxyAwareKafkaContainer) new ProxyAwareKafkaContainer()
+                                                     .withExposedPorts(TOXY_PROXY_PORT)
+                                                     .withNetwork(network);
+
+    public static ToxiproxyContainer toxiproxy  = new ToxiproxyContainer().withNetwork(network);
```
[the remark this answers is not in the document]
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/test/java/org/kie/server/springboot/samples/ProxyAwareKafkaEmitterTest.java:59]: done

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
@kiegroup/gatekeepers this is ready to merge (just samples to test Emitter Kafka integration with springboot and testcontainers).

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/main/java/org/kie/server/springboot/samples/kafka/WebSecurityConfig.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@Configuration("kieServerSecurity")
+@EnableWebSecurity
+public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http
+        .csrf().disable()
```
Comment (mareknovotny) [inline kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/main/java/org/kie/server/springboot/samples/kafka/WebSecurityConfig.java:36]: is this good to show not recommended CSRF disabling? see [sonar complaining about it](https://sonarcloud.io/code?id=org.drools%3Adroolsjbpm-integration&pullRequest=2313&selected=org.drools%3Adroolsjbpm-integration%3Akie-spring-boot%2Fkie-spring-boot-samples%2Fkie-server-spring-boot-kafka-sample%2Fsrc%2Fmain%2Fjava%2Forg%2Fkie%2Fserver%2Fspringboot%2Fsamples%2Fkafka%2FWebSecurityConfig.java&line=36)

**Comment by gmunozfe** [comment]
jenkins retest this

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/main/java/org/kie/server/springboot/samples/kafka/WebSecurityConfig.java (hunk 17 above)
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-kafka-sample/src/main/java/org/kie/server/springboot/samples/kafka/WebSecurityConfig.java]: @mareknovotny I think for this sample it doesn't matter (no production code and tests are not related to auth), I added the annotation _//NOSONAR_ for skipping the analysis in just that line, do you agree?

**Pull request merged**